src/test/app/AMMClawback_test.cpp

```diff
@@ -0,0 +1,1449 @@
+            USD(4500),
+            STAmount(EUR, UINT64_C(9000000000000001), -12),
+            IOUAmount{6363961030678928, -12}));
+
+        BEAST_EXPECT(amm.expectLPTokens(gw, IOUAmount{7071067811865480, -13}));
+        BEAST_EXPECT(amm.expectLPTokens(gw2, IOUAmount{1414213562373095, -12}));
+        BEAST_EXPECT(
+            amm.expectLPTokens(alice, IOUAmount{4242640687119285, -12}));
+
+        BEAST_EXPECT(env.balance(alice, USD) == USD(2000));
+        BEAST_EXPECT(env.balance(alice, EUR) == EUR(0));
+        BEAST_EXPECT(env.balance(gw, EUR) == EUR(4000));
+        BEAST_EXPECT(env.balance(gw2, USD) == USD(3000));
+
+        // gw2 claws back 4000 EUR from alice.
+        env(amm::ammClawback(gw2, alice, EUR, USD, EUR(4000), std::nullopt),
+            ter(tesSUCCESS));
+        env.close();
+        BEAST_EXPECT(amm.expectBalances(
+            USD(2500),
+            STAmount(EUR, UINT64_C(5000000000000001), -12),
+            IOUAmount{3535533905932738, -12}));
+
+        BEAST_EXPECT(amm.expectLPTokens(gw, IOUAmount{7071067811865480, -13}));
+        BEAST_EXPECT(amm.expectLPTokens(gw2, IOUAmount{1414213562373095, -12}));
+        BEAST_EXPECT(
+            amm.expectLPTokens(alice, IOUAmount{1414213562373095, -12}));
+
+        BEAST_EXPECT(env.balance(alice, USD) == USD(4000));
+        BEAST_EXPECT(env.balance(alice, EUR) == EUR(0));
+        BEAST_EXPECT(env.balance(gw, EUR) == EUR(4000));
+        BEAST_EXPECT(env.balance(gw2, USD) == USD(3000));
+    }
+
+    void
+    testNotHoldingLptoken(FeatureBitset features)
+    {
+        testcase(
+            "test AMMClawback from account which does not own any lptoken in "
+            "the pool");
+        using namespace jtx;
+
+        Env env(*this, features);
+        Account gw{"gateway"};
+        Account alice{"alice"};
+        env.fund(XRP(1000000), gw, alice);
+        env.close();
+
+        // gw sets asfAllowTrustLineClawback.
+        env(fset(gw, asfAllowTrustLineClawback));
+        env.close();
+        env.require(flags(gw, asfAllowTrustLineClawback));
+
+        auto const USD = gw["USD"];
+        env.trust(USD(100000), alice);
+        env(pay(gw, alice, USD(5000)));
+
+        AMM amm(env, gw, USD(1000), XRP(2000), ter(tesSUCCESS));
+        env.close();
+
+        // Alice did not deposit in the amm pool. So AMMClawback from Alice
+        // will fail.
+        env(amm::ammClawback(gw, alice, USD, XRP, USD(1000), std::nullopt),
+            ter(tecINTERNAL));
+    }
+
+public:
+    void
+    run() override
+    {
+        FeatureBitset const all{jtx::supported_amendments()};
+        testInvalidRequest(all);
+        testFeatureDisabled(all - featureAMMClawback);
+        testAMMClawbackSpecificAmount(all);
+        testAMMClawbackExceedBalance(all);
+        testAMMClawbackAll(all);
+        testAMMClawbackSameIssuerAssets(all);
+        testAMMClawbackSameCurrency(all);
+        testAMMClawbackIssuesEachOther(all);
+        testNotHoldingLptoken(all);
```

Comment:
There are no tests when the asset/s are frozen. We should have tests where the asset was first globally/individually frozen and then clawed back.

Reply:
We should also include a test where 3 holders deposit, then global freeze, then clawed back one by one. (Might be overkill)

Reply:
added in the recent commit